src/backend/access/appendonly/aomd.c

```diff
@@ -250,6 +251,9 @@ mdunlink_ao_perFile(const int segno, void *ctx)
 	char *segPathSuffixPosition = unlinkFiles->segpathSuffixPosition;
 
 	sprintf(segPathSuffixPosition, ".%u", segno);
+
+	do_truncate(segPath);
```

Comment:
I'm just curious, I didn't dig deep into the code. You've written in your ticket that heap tables are not affected by that issue: "space might be not freed right after table drop, because files are captured by writer process (seems quit long). Space is freed some time later (even without session end)."   Could you please extend this notice, does the space is freed by the same do_trancate call from the [patch](https://github.com/greenplum-db/gpdb/pull/11988)? Why there is a delay in space freeing? What are the steps between DROP TABLE call and freeing up the space? And why there is no such delay for AO tables? Just biefly. Thanks

Reply:
The issue you are talking about is reproduced only for TEMP heap tables. 
To reproduce it, you need to add some significant amount of data to temp relation and drop the relation. After that, writer process will be holding the descriptors for several minutes. After several minutes, without any additional steps, it will release the file descriptor.

I was able to reproduce it in the auto test too, by commands:
```
CREATE TEMP TABLE t(a int, b int) TABLESPACE testspace;
INSERT INTO t SELECT generate_series(1,20000000), generate_series(1,20000000);
DROP TABLE t;
\! find /proc/[0-9]*/fd -lname '@testtablespace@*(deleted)' -exec stat -Lc '%s' {} \; 2>/dev/null | awk 'BEGIN {sum=0} {sum += $1} END {print "size:",  sum}';
```
but it requires more data to add to the table (it is not reproducable on my side if I insert significantly less data). If everybody is Ok with such test case, I will add it to PR.

For TEMP tables, do_truncate is skipped in mdunlinkfork:
```
...
		if (!RelFileNodeBackendIsTemp(rnode))
		{
			/* Prevent other backends' fds from holding on to the disk space */
			ret = do_truncate(path);
		}
		else
			ret = 0;
...
```
I'm not sure why temp relations were skipped in original commit for do_truncate. I will investigate a bit more this question.  Maybe it is a good time now to fix it, if there is no special reason to treat temp relations in this specific way.

Reply:
> INSERT INTO t SELECT generate_series(1,20000000), generate_series(1,20000000);

why not
```sql
INSERT INTO t SELECT i, i FROM generate_series(1,20000000) i;
```
?

Reply:
That was a 'quick and dirty' test, for final version surely can use your proposal.

Reply:
> If everybody is Ok with such test case, I will add it to PR.
I don't think that we need to add the heap table test case.

>After that, writer process will be holding the descriptors for several minutes
In means, that at DROP on a temp heap table we unlink the files without truncating, but writer still holds them and then releases. And who is holding descriptors for the AO tables? Insertion procedure? (after mdcreate_ao).
And why the issue arise only for AO (not necessary temp) tables created at custom **tablespace**? (I couldn't reproduce with default tablespace)

Reply:
> And who is holding descriptors for the AO tables? 
> Backends who executed insertion procedure? (after mdcreate_ao). 

Yes, for more details please refer to https://tracker.yandex.ru/ADBDEV-4566#658a679b54110211fa23a7d7
The backends, that called mdcreate_ao, are holding descriptors.

> For heap tables backends close those descriptors?

For heap tables, descriptors obtained in mdcreate are stored in SMgrRelation structure:
```
reln->md_fd[forkNum]->mdfd_vfd = fd;
```
and later these descriptors might be reused in mdopen, and then properly closed.

> And why the issue arise only for AO (not necessary temp) tables created at custom tablespace? 
> (I couldn't reproduce with default tablespace)

Hmm, strange, I can reproduce the problem by commands below even without custom tablespace:
```
create table t (a int) with (appendonly=true);
insert into t select generate_series(1,2000);
drop table t;
```
Right after these commands execution check open file descriptors via lsof (either in psql or in separate shell (but in this case do not terminate psql)):
```
postgres=# \! lsof +L1 | grep postgres
postgres  561013 reskin   18u   REG    8,3        0     0 2139329 /usr/re/share/ws8-ADBDEV-4565/data/dbfast1/demoDataDir0/base/12812/16393.1 (deleted)
postgres  561014 reskin   18u   REG    8,3        0     0 2139330 /usr/re/share/ws8-ADBDEV-4565/data/dbfast2/demoDataDir1/base/12812/16393.1 (deleted)
postgres  561015 reskin   18u   REG    8,3        0     0 2139328 /usr/re/share/ws8-ADBDEV-4565/data/dbfast3/demoDataDir2/base/12812/16393.1 (deleted)
```


Regarding writer process and heap temp tables - the writer process releases the descriptors after next checkpoint - refer to code in BackgroundWriterMain:
```
if (FirstCallSinceLastCheckpoint())
{
	/*
	 * After any checkpoint, close all smgr files.  This is so we
	 * won't hang onto smgr references to deleted files indefinitely.
	 */
	smgrcloseall();
}
```
and to code in CheckpointerMain
```
/*
 * After any checkpoint, close all smgr files.  This is so we
 * won't hang onto smgr references to deleted files indefinitely.
 */
smgrcloseall();
```
Timed checkpoint is done after CheckPointTimeout seconds, which is by default 300 seconds. That explains why we see that writer process releases files after several minutes after table drop.